### ISSUE TITLE
Search Filtering Option 2 exited on signal 11

### ISSUE THREAD

**Issue by aaronsinclair** [e2guardian/e2guardian#95]
Hi. 
e2guardian gives me two options on search filtering.. Option 1 works fine (banned and override search terms) but option 2 (banned, override and search weighting) fails with a signal 11 exit. See attached debug log.

Here are the relevant configuration files: 

```
searchregexplist = '/usr/local/etc/e2guardian/lists/searchregexplist'

#
# Banned Search Term list(s) for option 1
#bannedsearchlist = '/usr/local/etc/e2guardian/lists/bannedsearchlist'
#bannedsearchoveridelist = '/usr/local/etc/e2guardian/lists/bannedsearchoveridelist'

# Search term phrase lists (for Option 2)
searchtermlimit = 5
bannedsearchtermlist = '/usr/local/etc/e2guardian/lists/bannedsearchtermlist1'
weightedsearchtermlist = '/usr/local/etc/e2guardian/lists/weightedsearchtermlist'
exceptionsearchtermlist = '/usr/local/etc/e2guardian/lists/exceptionsearchtermlist'

```
.# cat /usr/local/etc/e2guardian/lists/bannedsearchtermlist1
```
<angryman>
```

.# cat /usr/local/etc/e2guardian/lists/weightedsearchtermlist
```
<asdf><10>
<asdftest><20>
```

.# cat /usr/local/etc/e2guardian/lists/exceptionsearchtermlist
```
middlesex
```

.# cat /usr/local/etc/e2guardian/lists/searchregexplist
```
# This defines sites which will be search term checked
#
# The format is: "extended regular expression"->"search words"
#
#   left is the site matching pattern
#   right is the search words


#YOUTUBE
"^http[s]?://[0-9a-z]+.youtube\.[a-z]+[-/%.0-9a-z]*\?search_query=([^&]*)\&.*"->"\1"
"^http[s]?://[0-9a-z]+.youtube\.[a-z]+[-/%.0-9a-z]*\?search_query=([^&]*)"->"\1"
"^http[s]?://[0-9a-z]+\.google\.[a-z]+[-/%.?_&=#0-9a-z]*[&\?]?q=([^&]*)\&.*"->"\1"


#GOOGLE
"^http[s]?://[0-9a-z]+\.google\.[a-z]+[-/%.?_=#0-9a-z]*&q=([^&]*)"->"\1"
"^http[s]?://[0-9a-z]+\.google\.[a-z]+[-/%.?_&=#0-9a-z]*&q=([^&]*)\&.*"->"\1"

#BING
"^http[s]?://[0-9a-z]+\.bing\.[a-z]+[-/%.?_&=#0-9a-z]*?q=([^&]*)\&.*"->"\1"

 # BING SUGGESTIONS
"^http[s]?://[0-9a-z]+\.bing\.[a-z]+[-/%.?_&=#0-9a-z]*&qry=([^&]*).*"->"\1"
"^http[s]?://[0-9a-z]+\.bing\.[-/%.?_&=#0-9a-zA-Z]*&qry=([^&]*).*"->"\1"

#TWITTER
"^http[s]?://[0-9a-z]*[\.]?twitter\.[-/%.?_&=#0-9a-zA-Z]*[&?]q=([^&]*).*"->"\1"
"^http[s]?://[0-9a-z]*[\.]?twitter\.[-/%.?_&=#0-9a-zA-Z]*&query=([^&]*).*"->"\1"

#YAHOO

"^http[s]?://[0-9a-z]*\.[0-9a-z]*\.yahoo.[0-9a-z]*/.*p=([a-z0-9%-._~]*)"->"\1"

```
----------------------------------------------------------------------------------------

[debuglogSearchCrash.txt](https://github.com/e2guardian/e2guardian/files/42467/debuglogSearchCrash.txt)

**Comment by fredbcode** [comment]
>  fails with a signal 11 exit

At start ?

**Comment by aaronsinclair** [comment]
Not at start. E2g will work as expected up until a searchregex match is made and then exit sig 11 as per the debug attached.

**Comment by philipianpearce** [comment]
To help diagnosis can you try with the following lines commented out? 

bannedsearchtermlist = '/usr/local/etc/e2guardian/lists/bannedsearchtermlist1' 
weightedsearchtermlist = '/usr/local/etc/e2guardian/lists/weightedsearchtermlist' 
exceptionsearchtermlist = '/usr/local/etc/e2guardian/lists/exceptionsearchtermlist' 

Should work on the main content checking lists.  This is the mode that I have working. 

Regards 

Philip 

----- Original Message -----

**Comment by aaronsinclair** [comment]
Ok, testing now and will get back to you in due course

**Comment by fredbcode** [comment]
Hum, maybe I made something wrong but nothing works for me
I'm trying with a basic configuration for bing (no SSL HTTP) and some keywords like e2guardian
 
"^http[s]?://[0-9a-z]+\.bing\.[a-z]+[-/%.?_&=#0-9a-z]*?q=([^&]*)\&.*"->"\1"
or 
"^http://[0-9a-z]+\.bing\.[a-z]+[-/%.?_&=#0-9a-z]*?q=([^&]*)\&.*"->"\1"

bannedsearchtermlist
Tried with <e2guardian> or e2guardian

And
searchtermlimit = 5

**Comment by fredbcode** [comment]
Do you have the same behavior with dev branch ?

**Comment by aaronsinclair** [comment]
Hi Fred. Tested with the dev branch and search term weighting is working correctly using a mix of both option 1 and option 2.. 

.# Banned Search Term list(s) for option 1
bannedsearchlist = '/usr/local/etc/e2guardian/lists/bannedsearchlist'
bannedsearchoveridelist = '/usr/local/etc/e2guardian/lists/bannedsearchoveridelist'

.# Search term limit (for Option 2)
searchtermlimit = 20

```
1448561666.051          172.16.1.101    172.16.1.101    aaron_s3.secure.local   https://www.google.co.nz/search?safe=strict&hl=en&noj=1&site=webhp&q=bras+butt+cum&oq=bras+butt+cum&gs_l=serp.3...1804.4145.0.4283.4.4.0.0.0.0.271.271.2-1.1.0....0...1c.1.64.serp..4.0.0.fCNxEFy6s-Y        GET     403     0       -       -       -       10      -       454     *DENIED-PARAMS* Weighted search term limit of 20 : 32 ( bras + butt + cum )     32      Pornography     RelaxedFilterMode    2
1448561677.433          172.16.1.101    172.16.1.101    aaron_s3.secure.local   https://www.google.co.nz/complete/search?client=serp&hl=en-NZ&gs_rn=64&gs_ri=serp&tok=BHNryNJWTplGY83UJkNJRw&pq=bras%20butt&cp=3&gs_id=c&q=sex&xhr=t    GET 403      0       -       -       -       7       -       521     *DENIED* Banned Search Words: sex       0       BannedSearchWords       RelaxedFilterMode       2
```

**Comment by fredbcode** [comment]
> Hi Fred. Tested with the dev branch and search term weighting is 
> working correctly using a mix of both option 1 and option 2..
>
> .# Banned Search Term list(s) for option 1
> bannedsearchlist = '/usr/local/etc/e2guardian/lists/bannedsearchlist'
> bannedsearchoveridelist = 
> '/usr/local/etc/e2guardian/lists/bannedsearchoveridelist'
>
> .# Search term limit (for Option 2)
> searchtermlimit = 20
>
> |1448561666.051 172.16.1.101 172.16.1.101 aaron_s3.secure.local 
> https://www.google.co.nz/search?safe=strict&hl=en&noj=1&site=webhp&q=bras+butt+cum&oq=bras+butt+cum&gs_l=serp.3...1804.4145.0.4283.4.4.0.0.0.0.271.271.2-1.1.0....0...1c.1.64.serp..4.0.0.fCNxEFy6s-Y 
> GET 403 0 - - - 10 - 454 *DENIED-PARAMS* Weighted search term limit of 
> 20 : 32 ( bras + butt + cum ) 32 Pornography RelaxedFilterMode 2 
> 1448561677.433 172.16.1.101 172.16.1.101 aaron_s3.secure.local 
> https://www.google.co.nz/complete/search?client=serp&hl=en-NZ&gs_rn=64&gs_ri=serp&tok=BHNryNJWTplGY83UJkNJRw&pq=bras%20butt&cp=3&gs_id=c&q=sex&xhr=t 
> GET 403 0 - - - 7 - 521 *DENIED* Banned Search Words: sex 0 
> BannedSearchWords RelaxedFilterMode 2 |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/e2guardian/e2guardian/issues/95#issuecomment-159971695>.
>

Hi, I made some changes but something still wrong, no crash now ?
Please can you make a test without SSLMITM, a fake website should be 
denied, something like http://intranet.m/|search?=badword|
Thanks for your help and report

**Comment by aaronsinclair** [comment]
HI. 
Compiled from latest devel branch.. 
Set **MITM to  OFF** in e2guardianf1.conf 
- for HTTP site..  Expected outcome
```
1448565831.871          172.16.1.101    172.16.1.101    aaron_s3.secure.local   http://www.bing.com/search?q=adultos+bras&qs=n&form=QBLH&pq=adultos+bras&sc=8-8&sp=-1&sk=&cvid=91DCE292B5F247A4A4B8B63680AD9B5E GET     403     0       -   --       34      -       454     *DENIED-PARAMS* Weighted search term limit of 40 : 42 ( adultos+ bras ) 42      Pornography     RelaxedFilterMode       2
```

Tested with HTTPS site... no filtering && search term allowed--- expected outcome..


-------------------------------------------------------------------------

Set **MITM  to  ON** in e2guardianf1.conf 
tested http and https -- both working.
``` 
1448565902.617          172.16.1.101    172.16.1.101    aaron_s3.secure.local   http://www.bing.com/search?q=adultos+bras&qs=n&form=QBLH&pq=adultos+bras&sc=0-0&sp=-1&sk=&cvid=91DCE292B5F247A4A4B8B63680AD9B5E GET     403     0       -   --       31      -       454     *DENIED-PARAMS* Weighted search term limit of 40 : 42 ( adultos+ bras ) 42      Pornography     RelaxedFilterMode       2

1448566186.342          172.16.1.101    172.16.1.101    aaron_s3.secure.local   https://www.bing.com/search?q=adultos+bras&qs=n&form=QBLH&pq=adultos+bras&sc=8-5&sp=-1&sk=&cvid=2B7F8A4C74324352B4CEB008048815C4        GET     403     0   --       -       13      -       454     *DENIED-PARAMS* Weighted search term limit of 40 : 42 ( adultos+ bras ) 42      Pornography     RelaxedFilterMode       2
```

Setup a  searchregexplist regex that would catch search term in http://intranet.m/?search=badword
```
#TEST
"^http[s]?://intranet\.[a-z]+[-/%.0-9a-z]*\?search=([^&]*)"->"\1"
```

Tested for fake website http://intranet.m/?search=adultos+bras  and was blocked as expected

```
1448566744.857          172.16.1.101    172.16.1.101    aaron_s3.secure.local   http://intranet.m/?search=adultos+bras  GET     403     0       -       -       -       7       -       454     *DENIED-PARAMS* Weighted search term limit of 40 : 42 ( adultos+ bras )      42      Pornography     RelaxedFilterMode       2
```

**Comment by aaronsinclair** [comment]
I did some more testing on this to give a summary.. 
Configuration with #bannedsearchlist  and #bannedsearchoveridelist both commented out exited on signal 11
```
searchregexplist = '/usr/local/etc/e2guardian/lists/searchregexplist'
#
# Banned Search Term list(s) for option 1
#bannedsearchlist = '/usr/local/etc/e2guardian/lists/bannedsearchlist'
#bannedsearchoveridelist = '/usr/local/etc/e2guardian/lists/bannedsearchoveridelist'

# Search term phrase lists (for Option 2)
searchtermlimit = 5
bannedsearchtermlist = '/usr/local/etc/e2guardian/lists/bannedsearchtermlist1'
weightedsearchtermlist = '/usr/local/etc/e2guardian/lists/weightedsearchtermlist'
exceptionsearchtermlist = '/usr/local/etc/e2guardian/lists/exceptionsearchtermlist'
```
**However the next config worked as expected**
```
searchregexplist = '/usr/local/etc/e2guardian/lists/searchregexplist'
#
# Banned Search Term list(s) for option 1
bannedsearchlist = '/usr/local/etc/e2guardian/lists/bannedsearchlist'
bannedsearchoveridelist = '/usr/local/etc/e2guardian/lists/bannedsearchoveridelist'

# Search term phrase lists (for Option 2)
searchtermlimit = 5
bannedsearchtermlist = '/usr/local/etc/e2guardian/lists/bannedsearchtermlist1'
weightedsearchtermlist = '/usr/local/etc/e2guardian/lists/weightedsearchtermlist'
exceptionsearchtermlist = '/usr/local/etc/e2guardian/lists/exceptionsearchtermlist'
```

**Comment by fredbcode** [comment]
I can't reproduce :(
Can you share your files, please ?

**Comment by philipianpearce** [comment]
Fixed in develop branch

**Comment by fredbcode** [comment]
Should be fixed now, with 3.4.0.2
